metricbeat/module/haproxy/haproxy.go

```diff
@@ -137,60 +139,37 @@ type Info struct {
 }
 
 // Client is an instance of the HAProxy client
+type clientProto interface {
+	Stat() (*bytes.Buffer, error)
+	Info() (*bytes.Buffer, error)
+}
+
 type Client struct {
-	Address     string
-	ProtoScheme string
+	proto clientProto
 }
 
 // NewHaproxyClient returns a new instance of HaproxyClient
 func NewHaproxyClient(address string) (*Client, error) {
-	parts := strings.Split(address, "://")
-	if len(parts) != 2 {
-		return nil, errors.New("must have protocol scheme and address")
-	}
-
-	if parts[0] != "tcp" && parts[0] != "unix" {
-		return nil, errors.New("invalid protocol scheme")
-	}
-
-	return &Client{
-		Address:     parts[1],
-		ProtoScheme: parts[0],
-	}, nil
-}
-
-// Run sends a designated command to the haproxy stats socket
-func (c *Client) run(cmd string) (*bytes.Buffer, error) {
-	var conn net.Conn
-	response := bytes.NewBuffer(nil)
-
-	conn, err := net.Dial(c.ProtoScheme, c.Address)
-	if err != nil {
-		return response, err
-	}
-
-	defer conn.Close()
-
-	_, err = conn.Write([]byte(cmd + "\n"))
-	if err != nil {
-		return response, err
-	}
-
-	_, err = io.Copy(response, conn)
+	u, err := url.Parse(address)
 	if err != nil {
-		return response, err
+		return nil, errors.New("invalid url")
 	}
 
-	if strings.HasPrefix(response.String(), "Unknown command") {
-		return response, fmt.Errorf("unknown command: %s", cmd)
+	switch u.Scheme {
+	case "tcp":
+		return &Client{&unixProto{Network: u.Scheme, Address: u.Host}}, nil
+	case "unix":
+		return &Client{&unixProto{Network: u.Scheme, Address: u.Path}}, nil
```

Comment:
I have modified this, before the url was parsed just splitting protocol and address, but now after the change to use url package for url parsing the address is `Host` for tcp, but `Path` for unix.

In any case, even if supported, there was no test coverage for unix socket, this will have to be revisited, as a volume in the docker would be needed to test this.

Reply:
Exactly :D